src/stactools/core/utils/raster_footprint.py

```diff
@@ -19,76 +22,660 @@
+    two successive points in the list are separated by 10 units and a
+    ``distance`` of 2 is provided, 4 new points will be added between the two
+    original points (one new point every 2 units of ``distance``).
+
+    Derived from code found at
+    https://stackoverflow.com/questions/64995977/generating-equidistance-points-along-the-boundary-of-a-polygon-but-cw-ccw # noqa
+
+    Args:
+        point_list (List[Tuple[float, float]]): The list of points to be
+            densified.
+        distance (float): The interval at which to insert additional points. A
+            smaller densification distance should be used when reprojection
+            results in greater curvature from the original geometry.
+
+    Returns:
+        List[Tuple[float, float]]: A list of the densified points.
+    """
+    points: Any = np.asarray(point_list)
+    dxdy = points[1:, :] - points[:-1, :]
+    segment_lengths = np.sqrt(np.sum(np.square(dxdy), axis=1))
+    total_length = np.sum(segment_lengths)
+    cum_segment_lengths = np.cumsum(segment_lengths)
+    cum_segment_lengths = np.insert(cum_segment_lengths, 0, [0])
+    cum_interp_lengths = np.arange(0, total_length, distance)
+    cum_interp_lengths = np.append(cum_interp_lengths, [total_length])
+    interp_x = np.interp(cum_interp_lengths, cum_segment_lengths, points[:, 0])
+    interp_y = np.interp(cum_interp_lengths, cum_segment_lengths, points[:, 1])
+    return [(x, y) for x, y in zip(interp_x, interp_y)]
+
+
+class RasterFootprint:
+    """An object for creating a convex hull polygon around all areas within an
+    raster that have data values (i.e., they do not have the nodata value).
+    This convex hull is termed the "footprint" of the raster data and is
+    returned by the :py:meth:`footprint` method as a polygon in a GeoJSON
+    dictionary for use as the geometry attribute of a STAC Item.
+
+    Two important operations during this calculation are the densification of
+    the footprint in the native CRS and simplification of the footprint after
+    reprojection to EPSG 4326. If the initial low-vertex polygon in the native
+    CRS is not densified, this can result in a reprojected polygon that does not
+    accurately represent the data footprint. For example, a MODIS asset
+    represented by a rectangular 5 point Polygon in a sinusoidal projection will
+    reproject to a parallelogram in EPSG 4326, when it would be more accurately
+    represented by a polygon with two parallel sides and two curved sides. The
+    difference between these representations is greater the further away from
+    the meridian and equator the asset is located.
+
+    After reprojection to EPSG 4326, the footprint may have more points than
+    desired. This can be simplified to a polygon with fewer points that maintain
+    a maximum distance to the original geometry.
+
+    Args:
+        data_array (npt.NDArray[Any]): The raster data used for the
```

Comment:
```suggestion
        data_array (numpy.NDArray[Any]): The raster data used for the
```

The documentation doesn't show `import _ as _`, so we need to provide the full Python module/package name.

Reply:
>Do you think we should deprecate the free functions now? Or wait until a later time?

I was going to hold off until the changes have been properly exercised via the MODIS stactools package.

>Also, looks like rasterio 1.3.5.post1 changed how some subdatasets are named: https://github.com/stac-utils/stactools/actions/runs/4156577655/jobs/7190456348#step:5:32? That feels strange to me...

Yeah, not sure how to handle this.

Reply:
> Do you think we should deprecate the free functions now? Or wait until a later time?

Free functions that duplicate class methods are now marked as deprecated since version 0.4.4. Should we bump that to 0.5.0?

Reply:
>Also, looks like rasterio 1.3.5.post1 changed how some subdatasets are named

I modified the failing test to handle rasterio's old and new subdataset naming by stripping the leading underscores off the names (and sorting, but the new order doesn't bother me). The underscore removal in the test feels like a hack. If anything, we should look at stripping these leading underscores off in `cogify_subdatasets`, since we end up a[dding yet another underscore](https://github.com/stac-utils/stactools/blob/327985c01bfe739c01db29d3b72b4cc75ed8ef9a/src/stactools/core/utils/convert.py#L115) in the file name. 

Thoughts? Is the different underscoring a breaking change?

Reply:
>The documentation doesn't show import _ as _, so we need to provide the full Python module/package name.

This is fixed.